### PR TITLE
Resolve Mypy erorrs in `v3` branch

### PR DESCRIPTION
**Pull request by DahnJ**
Attempts to address https://github.com/zarr-developers/zarr-python/issues/1593

There are still two unsolved issues around the nonexistant `Store.from_path` in `core.py`.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/v3/store/core.py [hunk 1]
```diff
@@ -26,7 +26,8 @@ def __init__(self, store: Store, path: Optional[str] = None):
 
     @classmethod
     def from_path(cls, pth: Path) -> StorePath:
-        return cls(Store.from_path(pth))
+        # NOT SOLVED: This is instantiating an ABC + there is no from_path method
```
Comment (DahnJ) [inline src/zarr/v3/store/core.py:29]: Not solved here, as per comment. What subclass of `Store` should this use?

**Comment by joshmoore** [comment]
Thanks, @DahnJ. I've launched the workflows.

**Comment by DahnJ** [comment]
Thanks @joshmoore I tried to fix a failing test (which I think happened because `Any` was used in a Pydantic model), can you launch them again please?

**Comment by joshmoore** [comment]
On their way! 🏇🏽

File: src/zarr/v3/store/core.py [hunk 2]
```diff
@@ -76,7 +77,8 @@ def make_store_path(store_like: StoreLike) -> StorePath:
         try:
             from upath import UPath
 
-            return StorePath(Store.from_path(UPath(store_like)))
+            # NOT SOLVED: Similar here, ABC instantiation + no from_path method
```
[the remark this answers is not in the document]
Comment (jhamman) [reply src/zarr/v3/store/core.py:80]: See above.

**Review by jhamman — COMMENTED** [review]
Thanks @DahnJ for jumping in here! Really happy to see this contribution. 

Some guiding principles that may help the work here progress. First, I would prioritize getting Mypy to pass, even if it requires ignoring a few problematic lines. Then, once things are passing again (see #1649 where we disable the mypy checks in the v3 CI), then we can come back and fix any TODOs.

File: src/zarr/v3/group.py [hunk 3]
```diff
@@ -46,11 +46,11 @@ def to_bytes(self) -> Dict[str, bytes]:
             return {ZARR_JSON: json.dumps(self.to_dict()).encode()}
         else:
             return {
-                ZGROUP_JSON: self.zarr_format,
+                ZGROUP_JSON: str(self.zarr_format).encode(),
```
[the remark this answers is not in the document]
Comment (DahnJ) [reply src/zarr/v3/group.py]: Addressed in https://github.com/zarr-developers/zarr-python/pull/1692/commits/a30af88060b2db34717679e00c29cf2a03ddd265

**Comment by DahnJ** [comment]
@jhamman I have addressed the comments in https://github.com/zarr-developers/zarr-python/pull/1692/commits/a30af88060b2db34717679e00c29cf2a03ddd265

I was not sure how to handle the  [second](https://github.com/zarr-developers/zarr-python/pull/1692#discussion_r1518985869) `Store.from_path()` reference. What should happen if the `store_like` argument is of type `str`? I tried to make a potential suggestion based on a commented-out code in https://github.com/zarr-developers/zarr-python/pull/1692/commits/f57527a2f142941ebce09b924bae0933f1885a5e. Let me please know if this is the correct way of handling that case.

As for getting mypy to pass, it does pass for me locally and I didn't add any `type: ignore` lines. However, there already are a few in `zarr/v3` – I removed some more in https://github.com/zarr-developers/zarr-python/pull/1692/commits/e1fdd1ee231f134724f42b9072ea8e684407ec4b and then turned on mypy in pre-commit in https://github.com/zarr-developers/zarr-python/pull/1692/commits/db6da340ccafc0267d054a8002c3ee892483b724.

**Comment by jhamman** [comment]
@DahnJ - I think this is a great step forward. I want @d-v-b to review this but from my perspective, it could go in now, even with the outstanding store issue.

**Review by d-v-b — APPROVED** [review]

**Comment by d-v-b** [comment]
This looks good! There's some overlap with stuff I did over in #1743, but that's not a problem. I will merge and rebase #1743 on this.

**Comment by d-v-b** [comment]
thanks @DahnJ !

**Pull request merged**